### ISSUE TITLE
Attaching to a runnig gradle (Java) debug server works just sometimes

### ISSUE THREAD

**Issue by TheLeoP** [mfussenegger/nvim-dap#263]
### Debug adapter definition and debug configuration

Os: Windows 10
Nvim version: 0.5
java-debug version: 0.32.0
Other plugins used:
- [nvim-jdtls](https://github.com/mfussenegger/nvim-jdtls)
- [nvim-dap-ui](https://github.com/rcarriga/nvim-dap-ui)
- [one-small-step-for-vimkind](https://github.com/jbyuki/one-small-step-for-vimkind)

Java dap configuration and adapter:
```
local util = require('jdtls.util')

dap.adapters.java = function(callback)
  util.execute_command({command = 'vscode.java.startDebugSession'}, function(err0, port)
    assert(not err0, vim.inspect(err0))
    print("puerto:", port)
    callback({
      type = 'server';
      host = '127.0.0.1';
      port = port;
    })
  end)
end

dap.configurations.java = {
  {
    type = 'java',
    request = 'attach',
    name = "Java attach",
    hostName = "127.0.0.1",
    port = 5005
  },
}
```
nLua (debugger of lua embeded on nvim) dap configuration and adapter:
```
dap.adapters.nlua = function(callback, config)
  callback({ type = 'server', host = config.host, port = config.port or 5005 })
end

dap.configurations.lua = {
  {
    type = 'nlua',
    request = 'attach',
    name = "nlua attach",
    host = '127.0.0.1',
    port = function()
      local val = tonumber(vim.fn.input('Port: '))
      return val
    end,
  }
}

```
Attached files:
- sample gradle project
[prueba.zip](https://github.com/mfussenegger/nvim-dap/files/6952020/prueba.zip)
- logs
[logs.log](https://github.com/mfussenegger/nvim-dap/files/6952106/logs.log)

Additional note:
I also tried the default nvim-jdtls config for nvim-dap, but it didn't work either. Since, when I used [vimspector](https://github.com/puremourning/vimspector), I manage to copy part of the nvim-jdtls config for dap and use it to attach to a runnig gradle debug server, I tried to do that instead with nvim-dap, but maybe my problem has somethig to do with not knowing how to propperly use nvim-dap or nvim-jdtls since I'm  a pretty new user of these plugins.

### Steps to Reproduce

Sample gradle Java project structure:
```
prueba
|--- .gradle/
|--- .settings/
|--- app/
|     |---- .settings/
|     |---- bin/
|     |---- build/
|     |---- src/
|            |---- main/
|            |       |--- java/
|            |       |      |---- prueba/
|            |       |             |---- App.java
|            |       |---- resources/
|            |---- test/
|--- gradle/
|---- [other files]
```

Doesn't work
1. Open a terminal on prueba/app/ and run `gradle run --debug-jvm`
2. Open prueba/app/src/main/java/prueba/App.java with Neovim
3. Run `require('dap').continue()`

Does work
1. Open a terminal on prueba/app/ and run `gradle run --debug-jvm`
2. Open prueba/app/src/main/java/prueba/App.java with Neovim (instance A)
3. Open the plugin file nvim-dap/lua/dap/session.lua on another Neovim instance (instance B)
4. On intance B, go to line 771 and place a breakpoint with `require('dap').toggle_breakpoint()`
5. On instance A, run `require('osv').launch()`. It will return a port number.
6. On instance B run `require('dap').continue()`, then enter the port number given on point 5. This will start the debugging of the lua code executed on instance A.
7. On instance A, run `require('dap').continue()`. The execution will be stopped on the breakpoint set on point 4.
8. On instance B run `require('dap').continue()`, it will hit the same breakpoint again (for some reason this just happens sometimes)
9. On instance B run `require('dap').continue()` to continue the execution, this time there will not be any other breakpoints to hit.

### Expected Result

In both cases, to start the debugging of App.java by attaching to the gradle debug server.

### Actual Result

On the first case, a message saying `Debug adapter didn't respond. Either the adapter is slow (then wait and ignore this) or there is a problem with your adapter or java configuration. Check the logs for errors (:help dap.set_log_level)` is displayed. Waiting doesn't do anything, nvim-dap never attachs itself to the running debug server.

On the second case, nvim-dap does attachs itself to the running debug server and starts the debugging correctly. Sometimes (I guess it has something to do with the point 8.) gradle will output
```
Listening for transport dt_socket at address: 5005
ERROR: transport error 202: recv error: Connection reset by peer
Listening for transport dt_socket at address: 5005
Hello World!
```
`Hello World!` is the output of App.java when it runs correctly, but I'm not sure what `ERROR: transport error 202: recv error: Connection reset by peer` means. I know this has nothing to do with nvim-dap, but I figured it may be helpful to understand whats going wrong or what I'm donig wrong.

**Comment by mfussenegger** [comment]
Based on the log you shared it looks like the debug session worked? The session got intialized and it stopped on the breakpoint until you sent a `disconnect`?

That you try to debug nvim-dap with one-small-step-for-vimkind is confusing me a bit - is that related to your java debug problem, or was that just trying to figure something out?

**Comment by TheLeoP** [comment]
The debbug session wasn't working, so I tried debbuging it with one-small-step-for-vimkind to see what the problem was. When I did it, following the steps I mentioned, the debbug session did work. So that's what confused me.
On the logs you can see a first request with no response and then, when I debug Neovim looking up for what was going wrong, It worked as expected

**Comment by TheLeoP** [comment]
Also, I'm sorry if I don't sound polite or if I writte strangely, English isn't my first language. Thanks for yor help.

**Comment by mfussenegger** [comment]
Not getting a response to the `initialize` response could point to an issue with the debug adapter. It could be that it happens if you try to debug the application before jdtls is done loading or building your project.


I'd recommend to use nvim-jdtls for the `JdtCompile` command (or `require('jdtls').compile()` function) and to add the LSP status to your statusline. That way you get more insights into what is happening.

For example, if you watch the statusline here you can see the different build actions:

https://user-images.githubusercontent.com/38700/128755321-8cc54e44-60bd-4d8b-bd10-2518e44c6c5b.mp4

You can get the LSP status messages using the `vim.lsp.util.get_progress_messages` function.

**Comment by TheLeoP** [comment]
I don't think the problem is in jdtls since I'm not using its default dap config on this issue. But, for testing, I tried what you suggested:
- Not using jdtCompile

https://user-images.githubusercontent.com/53507599/128761841-93e35324-df6f-4ecd-a5f7-d45f1e83bf28.mp4


- Using jdtCompile

https://user-images.githubusercontent.com/53507599/128761908-2d8795b7-48e5-4eda-a88d-e8dc15732f55.mp4


- Using one-small-step-for-vimkind

https://user-images.githubusercontent.com/53507599/128761936-d4a139fe-2a66-44bd-8799-93ebf171d343.mp4

As you can see, even thought I don not use nvim-jdt for anything (beyond LSP), it always fully starts. Do you have any other idea why the debug session works while using osv but doesn't work normally?

**Comment by mfussenegger** [comment]
Is there anything in the lsp logs? I'm not sure how using one-small-step-for-vimkind would have any effect anything, other than changing the timing

**Comment by TheLeoP** [comment]
The lsp and dap logs don't show anything important, I think.

[dap.log](https://github.com/mfussenegger/nvim-dap/files/6957487/dap.log)

[lsp.log](https://github.com/mfussenegger/nvim-dap/files/6957488/lsp.log)

And, I guess the problem has something to do with the timmint, I just don't know why. I manage to find that, if I place the breakpoint on line 747 of session.lua, the debug session starts, but if I place the breakpoint on line 768 on the same file, the debug session does not start. Do you have any ide on why this could be happening?

```
function Session:request(command, arguments, callback)
  local payload = {
    seq = self.seq;
    type = 'request';
    command = command;
    arguments = arguments
  }
  log.debug('request', payload)
  local current_seq = self.seq
  self.seq = self.seq + 1
  vim.schedule(function()
    local msg = rpc.msg_with_content_length(vim.fn.json_encode(payload))
    self.client.write(msg)                                          [line 747]
    if callback then                                                   [line 748]
      self.message_callbacks[current_seq] = callback
      self.message_requests[current_seq] = arguments
    end
  end)
end
```

**Comment by mfussenegger** [comment]
Could you try with https://github.com/mfussenegger/nvim-dap/pull/265 ?

**Comment by TheLeoP** [comment]
Still the same problem. If I don't put a breakpoint on `self.client.write(msg)` (now line 751), the debbug session doesn't start. I even try adding an `print('foo')` on line 752 and putting a breakpoint on that line, but the debbug session doesn't start either.

I guess the timing problem could have something to do with `self.client.write(msg)`

**Comment by mfussenegger** [comment]
Might be that it already tries to send the initialize before the client is connected to the server. Could you try with https://github.com/mfussenegger/nvim-dap/pull/266 ?

**Comment by TheLeoP** [comment]
YES! That solves the problem, the debbug session now starts with both my attach config and with jdtls launch config. THANKS.

The message `Debug adapter didn't respond. Either the adapter is slow (then wait and ignore this) or there is a problem with your adapter or java configuration. Check the logs for errors (:help dap.set_log_level)` still shows up, but the debbugin works as expected.

**Comment by mfussenegger** [comment]
That would mean the debug adapter takes more than two seconds to initialize. 

I guess I could increase the time a bit, but I don't want to go too  high - otherwise users again get the impression that nothing is happening. Or I could add a setting for it

**Comment by TheLeoP** [comment]
I think a setting for it could be a solution, but, personally, I don't mind the message showing up. 
Also, once again  thanks for helping me with the bug
